File: apps/workflow-a/src/handlers/finalize.handler.ts
import type { Context } from 'aws-lambda';
import { createLogger } from '@org/shared-observability';
import { WORKFLOW_NAME, type PropostaEvent, type StepResult } from './types.js';

const logger = createLogger({ service: WORKFLOW_NAME, handler: 'finalize' });

/**
 * Terminal step: persists the final outcome and (in production) would emit a
 * "proposta.finalized" event back onto the shared topic for downstream
 * consumers. Here we just log and return the finalized result.
 */
export const handler = async (
  event: PropostaEvent,
  context: Context
): Promise<StepResult> => {
  const proposta = event.payload;

  logger.info('proposta.finalized', {
    requestId: context.awsRequestId,
    propostaId: proposta.id,
    outcome: proposta.status,
  });

  return {
    workflow: WORKFLOW_NAME,
    step: 'finalize',
    propostaId: proposta.id,
    status: 'FINALIZED',
    detail: { outcome: proposta.status },
  };
};

File: apps/workflow-a/Handlers/FinalizeHandler.cs
using Amazon.Lambda.Core;
using Acme.Shared.Domain;
using Acme.Shared.Observability;

namespace Acme.WorkflowA.Handlers;

/// <summary>
/// Terminal step: persists the final outcome and (in production) would emit a
/// "proposta.finalized" event back onto the shared topic for downstream
/// consumers. Here we log and return the finalized result.
/// </summary>
public sealed class FinalizeHandler
{
    private static readonly IStructuredLogger Logger = StructuredLogger.Create(
        new Dictionary<string, object?> { ["service"] = WorkflowConstants.Name, ["handler"] = "finalize" });

    public Task<StepResult> Handle(WorkflowEvent<Proposta> evt, ILambdaContext context)
    {
        var proposta = evt.Payload;

        Logger.Info("proposta.finalized", new Dictionary<string, object?>
        {
            ["requestId"] = context.AwsRequestId,
            ["propostaId"] = proposta.Id,
            ["outcome"] = proposta.Status.ToString(),
        });

        return Task.FromResult(new StepResult(
            WorkflowConstants.Name, "finalize", proposta.Id, Status.Finalized,
            new Dictionary<string, object?> { ["outcome"] = proposta.Status.ToString() }));
    }
}
